src/server/frontend_wayland/window_wl_surface_role.cpp

```diff
@@ -441,15 +441,20 @@ void mf::WindowWlSurfaceRole::commit(WlSurfaceState const& state)
     pending_explicit_height = std::experimental::nullopt;
 }
 
-void mf::WindowWlSurfaceRole::surface_destroyed()
+void mf::WindowWlSurfaceRole::surface_destroyed() noexcept
 {
     if (weak_client)
     {
         // "When a client wants to destroy a wl_surface, they must destroy this 'role object' wl_surface"
         // NOTE: the wl_shell_surface specification seems contradictory, so this method is overridden in it's implementation
-        BOOST_THROW_EXCEPTION(std::runtime_error{
-            "wl_surface@" + std::to_string(wl_resource_get_id(surface->resource)) +
-            " destroyed before associated role"});
+        log_warning(
+            "wl_surface@%d destroyed before associated role",
+            wl_resource_get_id(surface->resource));
+        // Post error manually instead of throwing exception because this method is called by the surface destructor
+        wayland::post_implementation_error(
```

Comment:
This is not an implementation error; this is a client protocol violation, right?